R/bootstrap.R

```diff
@@ -953,56 +959,155 @@ buildTabset <- function(tabs, ulClass, textFilter = NULL, id = NULL,
 buildTabItem <- function(index, tabsetId, foundSelected, tabs = NULL,
                          divTag = NULL, textFilter = NULL) {
 
-  divTag <- if (!is.null(divTag)) divTag else tabs[[index]]
+  divTag <- divTag %||% tabs[[index]]
 
   if (is.character(divTag) && !is.null(textFilter)) {
     # text item: pass it to the textFilter if it exists
     liTag <- textFilter(divTag)
     divTag <- NULL
 
-  } else if (inherits(divTag, "shiny.navbarmenu")) {
+  } else if (isTabPanelMenu(divTag)) {
     # navbarMenu item: build the child tabset
     tabset <- buildTabset(divTag$tabs, "dropdown-menu",
       navbarMenuTextFilter, foundSelected = foundSelected)
 
     # if this navbarMenu contains a selected item, mark it active
-    containsSelected <- containsSelectedTab(divTag$tabs)
-    liTag <- tags$li(
-      class = paste0("dropdown", if (containsSelected) " active"),
-      tags$a(href = "#",
-        class = "dropdown-toggle", `data-toggle` = "dropdown",
-        `data-value` = divTag$menuName,
-        getIcon(iconClass = divTag$iconClass),
-        divTag$title, tags$b(class = "caret")
-      ),
-      tabset$navList   # inner tabPanels items
-    )
+    title <- divTag$title
+    value <- divTag$menuName
+    icon <- getIcon(iconClass = divTag$iconClass)
+    active <- containsSelectedTab(divTag$tabs)
+    liTag <- tagFunction(function() {
+      if ("3" %in% getBootstrapVersion()) {
+        bs3NavItemDropdown(title, value, icon, active, tabset$navList)
+      } else {
+        # Bootstrap 4 is much more strict dropdown navs classes :eye-roll:
+        # Namely, the .dropdown-menu shouldn't contain .nav-link/.nav-item
+        # classes and should instead be a (top-level!) .dropdown-item
+        # (or else the .tab('show') method won't update the .active classes properly)
+        items <- tabset$navList
+        items$children <- lapply(items$children, function(x) {
+          # x should be a tagFunction() due to the else block below
+          x <- if (inherits(x, "shiny.tag.function")) x() else x
+          # Replace <li class="nav-item"><a class="nav-link"></a></li>
+          #    with <a class="dropdown-item"></a>
+          navItem <- x$children[[1]]
+          navItem$attribs$class <- "dropdown-item"
+          navItem
```

Comment:
The main unfortunate thing about all of this is that BS4 doesn't support `<li><a></a></li>` inside of dropdowns, which apparently they brought back in BS5

https://getbootstrap.com/docs/4.6/components/navs/#tabs-with-dropdowns
https://getbootstrap.com/docs/5.0/components/navs-tabs/#tabs-with-dropdowns

Might be worth asking for a backport into BS4, but I doubt we'll see it